rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+            CStr::from_ptr(message).to_str().unwrap().to_string()
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Debug)]
+pub struct Comment<'pr> {
+    comment: NonNull<yp_comment_t>,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Comment<'pr> {
+    /// Returns the text of the comment.
+    ///
+    /// # Panics
+    /// 
+    /// Panics if the comment is not able to be read.
+    /// 
+    #[must_use]
+    pub fn text(&self) -> String {
+        unsafe {
+            let start: *mut u8 = self.comment.as_ref().start as *mut u8;
+            let end: *mut u8 = self.comment.as_ref().end as *mut u8;
+
+            let len = end as usize - start as usize;
+            let slice: &[u8] = std::slice::from_raw_parts(start, len);
+
+            std::str::from_utf8(slice).unwrap().to_string()
+        }
+    }
+}
+
+/// A struct created by the `errors` or `warnings` methods on `ParseResult`. It
+/// can be used to iterate over the diagnostics in the parse result.
+pub struct Diagnostics<'pr> {
+    diagnostic: *mut yp_diagnostic_t,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Iterator for Diagnostics<'pr> {
+    type Item = Diagnostic<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.diagnostic.is_null() {
+            None
+        } else {
+            let current = Diagnostic {
+                diagnostic: unsafe { NonNull::new_unchecked(self.diagnostic) },
+                marker: PhantomData
+            };
+
+            self.diagnostic = unsafe { (*(self.diagnostic)).node.next.cast::<yp_diagnostic_t>() };
+            Some(current)
+        }
+    }
+}
+
+/// A struct created by the `comments` method on `ParseResult`. It can be used
+/// to iterate over the comments in the parse result.
+pub struct Comments<'pr> {
+    comment: *mut yp_comment_t,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Iterator for Comments<'pr> {
+    type Item = Comment<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.comment.is_null() {
+            None
+        } else {
+            let current = Comment {
+                comment: unsafe { NonNull::new_unchecked(self.comment) },
+                marker: PhantomData
+            };
+
+            self.comment = unsafe { (*(self.comment)).node.next.cast::<yp_comment_t>() };
+            Some(current)
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult<'pr> {
+    source: &'pr [u8],
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl<'pr> ParseResult<'pr> {
+    /// Returns the source string that was parsed.
+    #[must_use]
+    pub const fn source(&self) -> &'pr [u8] {
+        self.source
+    }
+
+    /// Returns a slice of the source string that was parsed using the given
+    /// location range.
+    #[must_use]
+    pub fn slice(&self, location: &Location<'pr>) -> &'pr [u8] {
+        let ptr = self.source.as_ptr() as usize;
```

Comment:
can this be reworked to avoid casting the pointer to `usize`? There should be a pointer offset API we can use instead.